masu/database/koku_database_access.py

```diff
@@ -153,3 +159,75 @@ def does_db_entry_exist(self):
             (Boolean): "True/False",
         """
         return bool(self._get_db_obj_query().first())
+
+    def add(self, use_savepoint=True, **kwargs):
+        """
+        Add a new row to this table.
+
+        Args:
+            use_savepoint (bool) whether a transaction savepoint should be used
+            kwargs (Dictionary): Fields containing table attributes.
+
+        Returns:
+            (Object): new model object
+
+        """
+        new_entry = self._table(**kwargs)
+        if use_savepoint:
+            self.savepoint(self._session.add, new_entry)
+        else:
+            self._session.add(new_entry)
+        return new_entry
+
+    def delete(self, obj=None, use_savepoint=True):
+        """
+        Delete our object from the database.
+
+        Args:
+            obj (object) model object to delete
+            use_savepoint (bool) whether a transaction savepoint should be used
+        Returns:
+            None
+        """
+        if obj:
+            deleteme = obj
+        else:
+            deleteme = self._obj
+
+        if use_savepoint:
+            self.savepoint(self._session.delete, deleteme)
+        else:
+            self._session.delete(deleteme)
+
+    def commit(self):
+        """
+        Commit pending database changes.
+
+        Args:
+            None
+        Returns:
+            None
+        """
+        self._session.commit()
+
+    def savepoint(self, func, *args, **kwargs):
```

Comment:
There is rollback logic in the `KokuDBAccess` when being used as a context manager.  Does `savepoint()` cover anything that the session rollback does not?  Maybe if we were not using a context manager for database accessors?

Reply:
See: https://docs.sqlalchemy.org/en/13/orm/session_transaction.html#using-savepoint and https://www.postgresql.org/docs/9.6/sql-savepoint.html

Each savepoint begins a new sub-transaction that can be committed or rolled back. So, it increases the granularity of where the commit/rollback happens.

Reply:
ok got it.  Yeah I purposefully didn't catch the exceptions in that case because I didn't want to mask any exceptions from our logs for cases other than removing providers while processing.  

I thought I moved all of the database accessors to context managers in the non-test code.

Reply:
That could be true. But I think there's an additional benefit in that we can create multiple savepoints, even within a single transaction scope.

So, the context manager will create a savepoint, then each call to `.add()` will create an additional savepoint for cases like this:

```
with MyAccessor as accessor:
    data = fetch_data_somehow()
    for item in data:
        accessor.add(item)
```

Previously, there was one savepoint created at the context manager scope. Now, there's additional savepoints within that scope for each `INSERT` that is within that transaction. So, if a single `INSERT` fails, we'll trap it and log it, but we don't need to rollback the whole transaction.